### PR TITLE
Imgyeong/course crud

### PR DESCRIPTION
**Pull request by ImgyeongLee**
## Related Issues

#14 #11 

## Changes

- New data schema -> Room
  - Our Room Schema
```TypeScript
model Room {
    id           String     @id @default(cuid())
    createdAt    DateTime   @default(now())
    name         String
    description  String
    courseId     String     @map("courseId")
    course       Course     @relation(fields: [courseId], references: [id])
    creatorId    String     @map("creatorId")
    creator      User       @relation(fields: [creatorId], references: [id], name: "CreatorOfRoom")
    User         User[]     @relation("UserInRoom")
}
```
- Updated data schema -> Course
  - Our New Course Schema
```TypeScript
model Course {
    id           String     @id @default(cuid())
    createdAt    DateTime   @default(now())
    subject      String
    code         Int        @default(0)
    title        String
    User         User[]
    Room         Room[]
}
```
Example: `subject` is like `CS` or `MTH` and `code` is the number like `290` or `162`
- (**EDITED**)  Support mutations for Course
  - createCourse
  - deleteCourse
- (**EDITED**)  Support mutations for Room
  - createRoom
  - deleteRoom
  - updateRoom
- Basic page structures of the page and study room page are provided
- Installed 100ms and the basic structure of the chat UI/UX is supported.
- Refactored the files (structures)

## Post features

- [ ] Create RoomCard for the RoomPage
  - [ ] Create authToken by using roomCode under this component
  - [ ] Implement handleSubmit and allow the user to join the chat room

### CODE REVIEW

File: apps/web/app/lib/actions.ts [hunk 1]
```diff
@@ -39,6 +39,46 @@ export async function authorizeUser(email: string, password: string) {
...
+  await prisma.course.create({
+    data: {
+      subject,
+      code,
+      title,
+    },
+  });
+
+  return { success: true };
+}
```
Comment (Nyumat) [inline apps/web/app/lib/actions.ts]: You should treat Server Actions as you would public-facing API endpoints, and ensure that the user is authorized to perform the action. 

So what you have would be:

```ts
export async function createCourse(data: CourseFields) {
  // dummy just to show usage
  const session = await getSession();
  if (!sessison) throw new Error("Unauthorized");

  const { subject, code, title } = data;
  await prisma.course.create({
    data: {
      subject,
      code,
      title,
    },
  });

  return { success: true };
}
```

Instead.

Read more about it [here.](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#authentication-and-authorization)

File: apps/web/app/lib/actions.ts [hunk 2]
```diff
@@ -39,6 +39,46 @@ export async function authorizeUser(email: string, password: string) {
...
+    orderBy: [{ subject: "asc" }, { code: "asc" }],
+  });
+  return courses;
+}
+
+export async function deleteCourse(id: string) {
+  await prisma.course.delete({ where: { id } });
+  revalidatePath("/platform/courses");
+}
+
```
Comment (Nyumat) [inline apps/web/app/lib/actions.ts]: Server actions emit `POST` requests, meaning these methods are uneccessary and a bad practice. 

When using Next.js Server Actions, the only thing that should be in these files (`actions.ts`) are **mutations**. 

[Server actions are for mutations, not data fetching.](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#authentication-and-authorization:~:text=data%20mutations%20in%20Next.js%20applications.) If you DO want to fetch data, do it directly in the server components or route handlers.

File: apps/web/app/platform/course-materials/page.tsx [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import {
+  getCourses,
+  getCourse,
+  getAllCourses,
+  createCourse,
+} from "@/app/lib/actions";
```
Comment (Nyumat) [inline apps/web/app/platform/course-materials/page.tsx]: What does `pnpm run build` give you with this? They're...unused, no?

But from my previous comment, this is where you'd put these methods instead. Additionally, you'd add security safeguards to make sure only **authorized users** can do these mutations/data fetching.

File: apps/web/app/platform/study-groups/view.tsx [hunk 4]
```diff
@@ -1,50 +1,48 @@
...
+          <DialogDescription>Add your course here.</DialogDescription>
+        </DialogHeader>
+        <div className="grid gap-4 py-4">
+          <div className="grid grid-cols-4 items-center gap-4">
+            <Label htmlFor="name" className="text-right">
+              Course Code
+            </Label>
+            <Input
+              id="name"
+              placeholder="ex) CS 161"
```
Comment (Nyumat) [inline apps/web/app/platform/study-groups/view.tsx]: nit: typo

File: apps/web/app/platform/study-groups/view.tsx [hunk 5]
```diff
@@ -1,50 +1,48 @@
...
+              className="col-span-3 placeholder-opacity-50"
+            />
```
Comment (Nyumat) [inline apps/web/app/platform/study-groups/view.tsx]: I'd look into [react-hook-form](https://ui.shadcn.com/docs/components/form) for the course creation forms!

**Review by Nyumat — CHANGES_REQUESTED** [review]
Minor stuff but definitely address these. 

It's a good habit to do things "the right way" even if the alternative "just works"

**Comment by ImgyeongLee** [comment]
@Nyumat I updated the PR. Let me know if I missed something or if I need to fix something.

**Review by Nyumat — APPROVED** [review]
image

LGTM! We may need to setup prettier to avoid diffs on the package.json, but this is great!

**Pull request merged**